### DISCUSSION TITLE
Conceptual Class Diagram of KERI

### DISCUSSION THREAD

**Discussion by edeykholt** [WebOfTrust/keripy#630] [General]
Attached is a UML class diagram showing the KERI concepts of Inception and Rotation Events.  This is a subset of a much larger model I'm working on.  This is intentionally not aligned with the current specification or implementation, but is intended to help new people accurately grasp the concepts of KERI. 
Not everyone speaks or likes UML, but it works to help me reach a clear understanding.
It might inform user experience designs and terminology in the future.
There's much that could be discussed around this diagram, and I'd welcome the opportunity to discuss here or live.
![KERI Class Diagram - Key State - Discussion Draft 2023-12-12](https://github.com/WebOfTrust/keripy/assets/8242469/a03aeb74-cb13-4071-b338-ee32e97465ff)

**Comment by edeykholt** [comment]
@SmithSamuelM FYI

**Comment by daidoji** [comment]
@edeykholt should "Weighted Public Key" not be "Key identifier"?  Or is there some kind of weighted public key derivation using a previous key pair that I don't remember?

**Reply by edeykholt** [reply]
@daidoji Thanks for the question!  I was attempting to model multisig with a thresholded set of fractionally weighted public keys for the current authority, and a  thresholded set of fractionally weighted hash-of-public-keys for the pre-rotated next authority (candidates).  Then, simpler examples for a single public key for current authority and hash-of-public-key for future authority is just a simple usage of that.   I did get tripped up around the concept of reserve rotation: https://trustoverip.github.io/tswg-keri-specification/draft-ssmith-keri.html#name-reserve-rotation-example
and that led me to introduce a more general term Key Identifier, used for either a public key or hash of a public key. In any case, I'll update the diagram after re-thinking this.  I'm still learning...

**Reply by daidoji** [reply]
@edeykholt yeah these diagrams are great, that's just the first thing that popped out at me.

**Comment by edeykholt** [comment]
Updated...
![KERI Class Diagram - Key State - Discussion Draft](https://github.com/WebOfTrust/keripy/assets/8242469/45a31e4f-79ac-4669-b724-16199bf9d5aa)
